inc/Engine/Optimization/RUCSS/Warmup/ResourceFetcher.php

```diff
@@ -0,0 +1,175 @@
+<?php
+declare(strict_types=1);
+
+namespace WP_Rocket\Engine\Optimization\RUCSS\Warmup;
+
+use WP_Rocket\Engine\Optimization\AbstractOptimization;
+use WP_Rocket\Engine\Optimization\AssetsLocalCache;
+use WP_Rocket\Logger\Logger;
+
+class ResourceFetcher extends AbstractOptimization {
+
+	/**
+	 * Resources array.
+	 *
+	 * @var array
+	 */
+	private $resources = [];
+
+	/**
+	 * Assets local cache instance
+	 *
+	 * @var AssetsLocalCache
+	 */
+	private $local_cache;
+
+	/**
+	 * Resource fetcher process instance.
+	 *
+	 * @var ResourceFetcherProcess
+	 */
+	private $process;
+
+	/**
+	 * Resource constructor.
+	 *
+	 * @param AssetsLocalCache       $local_cache Local cache instance.
+	 * @param ResourceFetcherProcess $process     Resource fetcher process instance.
+	 */
+	public function __construct( AssetsLocalCache $local_cache, ResourceFetcherProcess $process ) {
+		$this->local_cache = $local_cache;
+		$this->process     = $process;
+	}
+
+	/**
+	 * Handle Collecting resources and save them into the DB
+	 *
+	 * @param string $html Page HTML.
+	 */
+	public function handle( $html ) {
+		// Grab resources from page HTML.
+		$this->find_styles( $html );
+		$this->find_scripts( $html );
+
+		if ( empty( $this->resources ) ) {
+			return;
+		}
+
+		// Send resources to the background process to be saved into DB.
+		foreach ( $this->resources as $resource ) {
+			$this->process->push_to_queue( $resource );
+		}
+
+		$this->process->save()->dispatch();
+	}
+
+	/**
+	 * Find link styles on the page HTML.
+	 *
+	 * @param string $html Page HTML.
+	 */
+	private function find_styles( $html ) {
+		$links = $this->find( '<link(?:[^>]+[\s"\'])?href\s*=\s*[\'"]\s*(?<url>[^\'"\s]+)\s*?[\'"](?:[^>]+)?\/?>', $html );
```

Comment:
This regex works for `<link rel="stylesheet" href="assets/css/styles.min.css" media="all">`,  
but not for `<link href="assets/css/styles.min.css" media="all" rel="stylesheet">` (doesn't seem to catch links that do not have anything before the `href`). I think we need to account for any order of attributes here.

Reply:
I used this regex to catch all links and it works in both cases:
`<link\s+([^>]+[\s"\'])?href\s*=\s*[\'"]\s*?(?<url>[^\'"]+\.css(?:\?[^\'"]*)?)\s*?[\'"]([^>]+)?\/?>`

https://regex101.com/r/cwe74M/1/

Reply:
@crystinutzaa, this regex may be a solution to the need for a regex below... the check for a url ending with the `.css` extension may eliminate all the other unwanted links. I'm trying to think now of any other circumstances where a link would point to a .css file that we don't want to include. Any ideas?

Reply:
I adjusted the old regex a bit to be like
```
<link\s+(?:[^>]+[\s"\'])?href\s*=\s*[\'"]\s*(?<url>[^\'"\s]+)\s*?[\'"](?:[^>]+)?\/?>
```
so it can get both link schemas.

Reply:
@iCaspar I was thinking about `<link rel="preload" href="path/to/file.css"/>` 
This still needs to be bailout based on rel attribute

Reply:
@engahmeds3ed's revised regex is working for me. Going to mark this resolved. Agreed that we do need the second regex below.